lib/elixir/lib/enum.ex

```diff
@@ -3217,19 +3217,30 @@ defmodule Enum do
       iex> Enum.sum(1..10//2)
       25
 
+      iex> Enum.sum(1..10, & &1 * 2)
+      209
+
+      iex> Enum.sum([%{num: 1}, %{num: 2}], & &1.num)
+      3
+
+      iex> Enum.sum(%{num_2: 2, num_3: 3}, fn {_, v} -> v end)
+      5
   """
-  @spec sum(t) :: number
-  def sum(enumerable)
+  @spec sum(t, (element -> number)) :: number
+  def sum(enumerable, fun \\ fn item -> item end)
+
+  def sum(first..last//step, fun) do
```

Comment:
This is probably the bit I'm more concerned about, from my testing it seems like the approach of using the `fun` on the "edges" of the range and then _coercing_ it back into a range, seems like the way to go. But I'm open to feedback on this 🙌

Reply:
This will only work if the function is linear, which we cannot guarantee. For example, &:math.sin/1 wouldn’t work as callback.